File: python/transpose/transpose.py
def transpose(input_lines):
    if len(input_lines) == 0:
        return ''
    input_lines = input_lines.split('\n')
    max_length = max([len(line) for line in input_lines])
    return "\n".join([map_line(col, input_lines) for col in range(0, max_length)])

def map_line(col, lines):
    return "".join([calc_value(row, col, line, lines) for row, line in enumerate(lines)])

# Look ahead
def ahead_blank(row, col, lines):
    return all([col >= len(lines[p_row]) for p_row in range(row, len(lines))])

def calc_value(row, col, line, lines):
    if col < len(line):
        return line[col]
    elif ahead_blank(row, col, lines):
        return ''
    else:
        return ' '

File: haskell/transpose/src/Transpose.hs
module Transpose (transpose) where

transpose :: [String] -> [String]
transpose lines = transpose' lines

transpose' :: [String] -> [String]
transpose'  lines
  | all null lines = []
transpose' lines = [head line | line <- lines] : transpose' [tail line | line <- lines]

fillBlank :: [String] -> [String]
fillBlank lines =
  let maxLength = maximum (map length lines)
      fill line = if length line == maxLength then line else line ++ (replicate (maxLength - (length line)) ' ')
   in map fill lines
